src/app/api/manuscript.api.ts

```diff
@@ -76,29 +88,91 @@ export async function getManuscriptContent(id: string): Promise<Manuscript> {
+  const backendTranscations = changes.reduce(reduceHistory, {});
+  compressChanges(backendTranscations);
   return axios.post(manuscriptUrl(id) + '/changes', { changes: backendTranscations });
 }
 
-function makeChangesSeralizable(diff: ManuscriptDiff): Record<string, unknown> {
-  // filter out props that don't refer to a prosemirror transaction.
-  const [transcation = null] = Object.keys(diff)
-    .filter((key) => diff[key] instanceof Transaction)
-    .map((key) => {
-      const transcation = diff[key] as Transaction;
-      return {
-        path: key,
-        steps: transcation.steps
+export async function getManuscriptChanges(id: string): Promise<ManuscriptChangesResponse['changes']> {
+  const manuscriptChangesResponse = await axios.get<ManuscriptChangesResponse>(`./changes/${id}/changes.json`);
+  return manuscriptChangesResponse.data.changes;
+}
+
+function reduceHistory(
+  acc: Record<string, SerializableChanges>,
+  diff: ManuscriptDiff
+): Record<string, SerializableChanges> {
+  Object.keys(diff).forEach((path) => {
+    if (path === '_timestamp') {
+      return;
+    }
+
+    const type: SerializableChangeType = diff[path] instanceof Transaction ? 'steps' : 'object';
+    if (!acc[path]) {
+      acc[path] = {
+        path,
+        type,
+        timestamp: diff._timestamp
       };
-    });
-  if (transcation) {
-    return { ...transcation, timestamp: diff._timestamp };
+    }
+
+    if (diff[path] instanceof Transaction) {
+      acc[path].steps = (acc[path].steps || []).concat((diff[path] as Transaction).steps);
+    } else {
+      acc[path].object = diff[path] as SerializableObjectValue;
+    }
+  });
+  return acc;
+}
+
+function compressChanges(changes: Record<string, SerializableChanges>): Record<string, SerializableChanges> {
+  const mergeOptions = Object.keys(changes).reduce((mergeAcc: Record<string, string>, path: string) => {
+    const crumbs = path.split('.');
+
+    for (let i = 1; i < crumbs.length; i++) {
+      const changesPath = crumbs.slice(0, i);
+
+      if (has(changes, changesPath.join('.'))) {
+        mergeAcc[path] = changesPath.join('.');
+        break;
+      }
+    }
+    return mergeAcc;
+  }, {});
+
+  for (const [mergeSourcePath, mergeDestPath] of Object.entries(mergeOptions)) {
+    changes[mergeDestPath] = mergeChanges(changes, mergeSourcePath, mergeDestPath);
+    delete changes[mergeSourcePath];
   }
-  return transcation;
+
+  return changes;
 }
 
-export async function getManuscriptChanges(id: string): Promise<ManuscriptChangesResponse['changes']> {
-  const manuscriptChangesResponse = await axios.get<ManuscriptChangesResponse>(`./changes/${id}/changes.json`);
-  return manuscriptChangesResponse.data.changes;
+function mergeChanges(
+  changes: Record<string, SerializableChanges>,
+  sourcePath: string,
+  destPath: string
+): SerializableChanges {
+  const source = changes[sourcePath];
+  const dest = changes[destPath];
+
+  if (dest.timestamp > source.timestamp) {
```

Comment:
Not an issue, but we will need to consider that clients clocks are likely not going to be accurate. So if these timestamps are set by each individual client using their current system time, then it's likely that there could be ordering issues applying the merges. Let's not worry about this now though.

Reply:
not really an issue here. These timestamps all belong to clients local history of changes so it doesn't really matter we only use that to identify what comes first for the purpose of squashing.